pkg/thing/gateway.go

```diff
@@ -48,48 +49,64 @@ var ErrCOAPServerAlreadyStarted = errors.New("CoAP server has already been start
 
 // ThingGateway represents the Thing Gateway
 type ThingGateway struct {
-	Thing     defaultThing
-	authCache *tokencache.Cache
+	gatewayThing     Thing
+	authCache        *tokencache.Cache
+	callbackHandlers []callback.Handler
 	// coap server
 	coapServer *coap.Server
 	coapChan   chan error
 	address    net.Addr
+	// AM connection
+	amConnection client.Connection
+	amURL        string
+	realm        string
+	authTree     string
 }
 
 // NewThingGateway creates a new Thing Gateway
 func NewThingGateway(baseURL string, realm string, authTree string, handlers []callback.Handler) *ThingGateway {
 	return &ThingGateway{
-		Thing: defaultThing{
-			connection: &amConnection{baseURL: baseURL, realm: realm, authTree: authTree},
-			handlers:   handlers,
-		},
-		authCache: tokencache.New(5*time.Minute, 10*time.Minute),
+		authCache:        tokencache.New(5*time.Minute, 10*time.Minute),
+		amURL:            baseURL,
+		realm:            realm,
+		authTree:         authTree,
+		callbackHandlers: handlers,
 	}
 }
 
 // Initialise the Thing Gateway
 func (c *ThingGateway) Initialise() error {
-	err := c.Thing.connection.initialise()
+	amURL, err := url.Parse(c.amURL)
 	if err != nil {
 		return err
 	}
-	return c.Thing.authenticate()
-}
-
-// AuthenticateWith the given authentication tree
-func (c *ThingGateway) AuthenticateWith(tree string) *ThingGateway {
```

Comment:
So the only way to change which tree a Gateway uses to authenticate a Thing would be to recreate the Gateway? We would lose all the caches etc.

Reply:
Yes, but I don't think we ever need to change the tree in normal operation. For the test I'm going to change the tree in AM rather than changing the gateway.